### PR TITLE
[WIP] Experimental fixes

### PR DESCRIPTION
**Pull request by jkuchar**
This PR contains experimental fixes and idea proposals. Comments please.
- Mail API: fix of header name formating (access using magic methods) (BC break)
- ImapDriver: removed server prefix from folder listing (BC break) related to #10
  - [ ] TODO: This should be configurable instead.
- ImapDriver does not parse address fields into ContactList object due to very buggy implementation of buggy addr list parser implementation (BC break)
  - [ ] TODO: write paser for that or provide option if you want to parse it or not
  - BTW I do not think that it is good idea to have parsing of this level in IMAP driver. This should be on higher level - e.g. in Mail or not at all and provide extra service for that.
- Another thing that should be mentioned is that php_imap extension is very buggy and it behaves stupid in many cases. E.g. mail->move() will copy and delete original message. (read from mail server logs) This and other bugs can be solver by implementing standalone IMAP client that will be not dependent on php_imap extension. Only working library for this case looks to be http://dev.horde.org/imap_client/ which is 100% PHP implementation of IMAP (no extensions required)

This PR requires #13 to be merged first.

### CODE REVIEW

File: MailLibrary/Drivers/ImapDriver.php [hunk 1]
```diff
@@ -381,7 +385,7 @@ public function setFlag($mailId, $flag, $value)
 	 * @throws DriverException
 	 */
 	public function copyMail($mailId, $toMailbox) {
-		if(!imap_mail_copy($this->resource, $mailId, $this->server . $this->encodeMailboxName($toMailbox), CP_UID)) {
+		if(!imap_mail_copy($this->resource, $mailId, /*$this->server .*/ $this->encodeMailboxName($toMailbox), CP_UID)) {
```
Comment (jkuchar) [inline MailLibrary/Drivers/ImapDriver.php:395]: There should be some way how to configure this.

File: MailLibrary/Mail.php [hunk 2]
```diff
@@ -78,7 +81,8 @@ public function __construct(Connection $connection, Mailbox $mailbox, $id)
 	public function __isset($name)
 	{
 		$this->headers !== NULL || $this->initializeHeaders();
-		return isset($this->headers[$this->formatHeaderName($name)]);
+		$key = $this->normalizeHeaderName($this->lowerCamelCaseToHeaderName($name));
```
Comment (jkuchar) [inline MailLibrary/Mail.php:84]: This is proposal. Basic idea is that magic methods uses magic name translation into normalized header name and getHeader() is more accurate (less magic included).

**Comment by greeny** [comment]
> Mail API: fix of header name formating (access using magic methods) (BC break)

I was thinking about this, and I think the best solution is:
- parse headers but DONT change keys, save in array (duplicite headers if any - only last one is used)
- access using `->getHeader("X-Requested-With")` - access directly target header (no magic here)
- access using `->xRequestedWith` - change camelCase into kebab-case with uppercased first letters in each word

I think this should work without any problems, do you have other opinion?

> ImapDriver: removed server prefix from folder listing (BC break) related to #10 

This is problem I had not figured out yet, how to solve. I am planning to do little test for most common email providers and check, how does their IMAP server react to create/rename/delete mailbox and move/copy/delete mail from mailbox, with both prefixed and non-prefixed mailbox name. I know, that 4 years ago when I was writing this library, I did some testing for gmail and this worked, so I used it and let it be. But things maybe changed or you used different provider, so this needs to be tested.

Since it is BC break, this should be only in 3.3.0 or even 4.0 (which is already in development, slow, but it is)

> ImapDriver does not parse address fields into ContactList object due to very buggy implementation of buggy addr list parser implementation (BC break)
>  TODO: write paser for that or provide option if you want to parse it or not
> BTW I do not think that it is good idea to have parsing of this level in IMAP driver. This should be on higher level - e.g. in Mail or not at all and provide extra service for that.

I completely agree this is not the best implementation. I wanted to provide easier access to parts of address, but did not think about it correctly (also somebody came with PR, which I merged without thinking, thats why there is some buggy code).

I think `getHeader()` should return raw header data, not parsed or anything. And I have to decide, if there will be some other method (e.g. `getContactHeader()`), which will provide parsed data, or if I will split this into some service.

The reason why it is in IMAP driver and not on higher level is because it uses some `imap_*` function for parsing. I am willing to write custom parser based on same RFC and split the functionality into some other service.

> Another thing that should be mentioned is that php_imap extension is very buggy and it behaves stupid in many cases. E.g. mail->move() will copy and delete original message. (read from mail server logs) This and other bugs can be solver by implementing standalone IMAP client that will be not dependent on php_imap extension. Only working library for this case looks to be http://dev.horde.org/imap_client/ which is 100% PHP implementation of IMAP (no extensions required)

That is part of refactoring in 4.X branch, which I will think about. Thanks for the link, I was looking some time ago for working library and did not find any. Maybe I can use that, or just write own client. It will be recommended driver, but I think I should not drop support for IMAP driver. Or should I?

File: MailLibrary/Drivers/ImapDriver.php (hunk 1 above)
Comment (greeny) [inline MailLibrary/Drivers/ImapDriver.php:395]: See my comment in PR

File: MailLibrary/Mail.php (hunk 2 above)
Comment (greeny) [inline MailLibrary/Mail.php:84]: See my comment in PR

**Comment by jkuchar** [comment]
> header: I think this should work without any problems, do you have other opinion?

No this will not work. What i come with works. parse header -> fix utf8 -> utf8 NFC normalization -> toLower -> save as key to array This is only what should works. E.g.: Cc, CC and cc is the same header

> server prefix

This should be configurable. And default to be turned on. So no BC break. I've been using it with Kerio Connect. 

Or there can be also auto-detection using mail listing. But this can be slow in big mailboxes.

> The reason why it is in IMAP driver and not on higher level is because it uses some imap_\* function for parsing. I am willing to write custom parser based on same RFC and split the functionality into some other service.

Cool!

> imap driver without imap extension
> I thing there should be MailClient library in one repo and drivers in another (because of dependencies). You you wish to continue developing driver for PHP imap extension you can but I do not see any advatages over using horde client.

**Pull request closed without merging**